misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/variable/types/BRecord.java

```diff
@@ -55,34 +64,67 @@ public Map<String, Value> computeChildVariables() {
+            }
+            return childVarMap;
         } catch (Exception ignored) {
             return new LinkedHashMap<>();
         }
     }
 
+    private Map<Value, Value> getRecordFields() {
+        try {
+            Optional<Method> getKeysMethod = VariableUtils.getMethod(jvmValue,
+                                                                    METHOD_GET_KEYS,
+                                                                    GETKEYS_METHOD_SIGNATURE_PATTERN);
+
+            Value keyArray = ((ObjectReference) jvmValue).invokeMethod(context.getOwningThread().getThreadReference(),
+                    getKeysMethod.get(), Collections.emptyList(), ObjectReference.INVOKE_SINGLE_THREADED);
+            loadedKeys = (ArrayReference) keyArray;
+
+            Map<Value, Value> recordFields = new LinkedHashMap<>();
+            List<Value> keysRange = loadedKeys.getValues(0, loadedKeys.length());
+
+            for (int i = 0; i < loadedKeys.length(); i++) {
+                Value key = keysRange.get(i);
+                recordFields.put(key, getValueFor(key));
+            }
+            return recordFields;
+        } catch (Exception ignored) {
+            return null;
+        }
+    }
+
+    private Value getValueFor(Value key) {
+        try {
+            Optional<Method> getMethod = VariableUtils.getMethod(jvmValue, METHOD_GET, GET_METHOD_SIGNATURE_PATTERN);
+            if (getMethod.isEmpty()) {
+                return null;
+            }
+            return ((ObjectReference) jvmValue).invokeMethod(context.getOwningThread().getThreadReference(),
+                    getMethod.get(), Collections.singletonList(key), ObjectReference.INVOKE_SINGLE_THREADED);
+        } catch (Exception ignored) {
+            return null;
+        }
+    }
+
     @Override
     public int getChildrenCount() {
         try {
             if (!(jvmValue instanceof ObjectReference)) {
                 return 0;
             }
-            ObjectReference jvmValueRef = (ObjectReference) jvmValue;
-            Map<Field, Value> fieldValueMap = jvmValueRef.getValues(jvmValueRef.referenceType().allFields());
-            long recordFieldCount = fieldValueMap.keySet()
-                    .stream()
-                    .filter(field -> field.toString().contains(RECORD_FIELD_PATTERN_IDENTIFIER))
-                    .count();
-
-            return Long.valueOf(recordFieldCount).intValue();
+            Optional<Method> getKeysMethod = VariableUtils.getMethod(jvmValue,
+                                                                    METHOD_GET_KEYS,
+                                                                    GETKEYS_METHOD_SIGNATURE_PATTERN);
+            Value keyArray = ((ObjectReference) jvmValue).invokeMethod(context.getOwningThread().getThreadReference(),
+                    getKeysMethod.get(), Collections.emptyList(), ObjectReference.INVOKE_SINGLE_THREADED);
+            return ((ArrayReference) keyArray).length();
```

Comment:
Can we store this `keyArray` in the `loadedKeys` field, so that we can load the keys only if the `loadedKeys` is null inside the `getRecordFields()` method? We can avoid retrieving the same keyset in multiple times in that way.

Reply:
or we can write a `loadAllKeys()` method and re-use it in both places